components/core/src/clp_s/JsonParser.cpp

```diff
@@ -557,6 +621,303 @@ int32_t JsonParser::add_metadata_field(std::string_view const field_name, NodeTy
+            }
+        case clp::ffi::SchemaTree::Node::Type::Obj:
+            if (node_has_value) {
+                if (node_value->is_null()) {
+                    return NodeType::NullValue;
+                }
+            }
+            return NodeType::Object;
+        default:
+            SPDLOG_ERROR("Unknown IR Node Type Detected");
+            throw OperationFailed(ErrorCodeFailure, __FILENAME__, __LINE__);
+    }
+}
+
+auto JsonParser::add_node_to_archive_and_translations(
+        uint32_t ir_node_id,
+        clp::ffi::SchemaTree::Node const& ir_node_to_add,
+        NodeType archive_node_type,
+        int32_t parent_node_id
+) -> int {
+    auto validated_escaped_key
+            = clp::ffi::validate_and_escape_utf8_string(ir_node_to_add.get_key_name());
+    std::string node_key;
+    if (validated_escaped_key.has_value()) {
+        node_key = validated_escaped_key.value();
+    } else {
+        SPDLOG_ERROR("Key is not UTF-8 compliant: \"{}\"", ir_node_to_add.get_key_name());
+        throw OperationFailed(ErrorCodeFailure, __FILENAME__, __LINE__);
+    }
+    int const curr_node_archive_id
+            = m_archive_writer->add_node(parent_node_id, archive_node_type, node_key);
+
+    m_ir_node_to_archive_node_id_mapping.emplace(
+            std::make_pair(ir_node_id, archive_node_type),
+            curr_node_archive_id
+    );
+    return curr_node_archive_id;
+}
+
+auto JsonParser::get_archive_node_id(
+        uint32_t ir_node_id,
+        NodeType archive_node_type,
+        clp::ffi::SchemaTree const& ir_tree
+) -> int {
+    int curr_node_archive_id{constants::cRootNodeId};
+    auto flat_map_location
+            = m_ir_node_to_archive_node_id_mapping.find(std::pair(ir_node_id, archive_node_type));
+
+    if (m_ir_node_to_archive_node_id_mapping.end() != flat_map_location) {
+        return flat_map_location->second;
+    }
+
+    std::vector<uint32_t> ir_id_stack;
+    ir_id_stack.push_back(ir_node_id);
+    int32_t next_parent_archive_id{constants::cRootNodeId};
+    NodeType next_node_type = archive_node_type;
+
+    while (true) {
+        auto const& curr_node = ir_tree.get_node(ir_id_stack.back());
+        auto parent_of_curr_node_id = curr_node.get_parent_id();
+        if (parent_of_curr_node_id.has_value()) {
+            ir_id_stack.push_back(parent_of_curr_node_id.value());
+            next_node_type = NodeType::Object;
+        } else {
+            next_parent_archive_id = constants::cRootNodeId;
+            break;
+        }
+
+        flat_map_location = m_ir_node_to_archive_node_id_mapping.find(
+                std::pair(ir_id_stack.back(), next_node_type)
+        );
+        if (m_ir_node_to_archive_node_id_mapping.end() != flat_map_location) {
+            curr_node_archive_id = flat_map_location->second;
+            next_parent_archive_id = flat_map_location->second;
+            ir_id_stack.pop_back();
+            break;
+        }
+    }
+
+    while (not ir_id_stack.empty()) {
```

Comment:
```suggestion
    while (false == ir_id_stack.empty()) {
```
I think our coding standard asks for explicit evaluation against false.